new spec

tmakar opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

tmakar commented:
@atmaxinger @markus2330 

This is a first draft of the "new" spec implementation. It contains already an implementation for copying meta data from `spec` to all other namespaces who have a matching key.

TODO:
- Support `_`
- Support array specification (`#`)

What @markus2330 and I agreed on:
- No `?` and `*`
- If array specification and wildcard both exist for the same key => error
- Assign / Condition , Conflicts => will not be implemented
- Validation (SET => if not correct, ERROR,  GET => if not correct, WARNING)
- tests for the new implementation (similar to `testmod_spec_old.c`)

I am bit confused about how elektra handles errors. May you @atmaxinger can clarify a little bit or show me a plugin where this error / warnings are done correctly.

Thank you for your time! 😉

tmakar commented:
The files with `old` implementation / tests are suffixed with `old`. Those files will be deleted before merging.

markus2330 commented:
> Assign / Condition , Conflicts => will not be implemented

assign/condition also should be removed from src/plugins/conditionals

> tests for the new implementation (similar to testmod_spec_old.c)

More and better tests are definitely needed.

atmaxinger commented:
@tmakar could you please name your new implementation `spec-new.c` or something like that? Currently, it is very hard to review as most lines are just marked as deleted.

tmakar commented:
@atmaxinger have renamed it.


Some questions to clarify (@atmaxinger and @markus2330):

1. Let's say we have a specification and in the hook stage the `spec` plugin comes in. It copies all the metadata from the spec namespace to the corresponding keys in the other namespace. 
It could be the case that there is a specification for a key which does not exist in any namespace (specification does not provide any metadata default nor require). What should happen in this case? Show a warning that there is a specification for a key that does not exist or anything else?

2. If I have a specification with a metakey `default` and the key does not exist in any of the namespaces, should it then automatically be created by the spec plugin?

3. If I have a specification with a metakey `require` and the key does not exist in any of the namespaces, should it then return an error or also create the key by itself?

In case 2. and 3. is answered with yes (both creating keys) should they be created in all namespaces?

Thanks for your time! :wink:

atmaxinger commented:
> May you @atmaxinger can clarify a little bit or show me a plugin where this error / warnings are done correctly.

AFAIK, it's convention that you add metakeys to the `parent` key. I.e. `meta:/error/spec/...`, `meta:/warning/spec/...`, `meta:/info/spec/...`

> It could be the case that there is a specification for a key which does not exist in any namespace (specification does not provide any metadata default nor require). What should happen in this case? Show a warning that there is a specification for a key that does not exist or anything else?

IMHO if the key is not required then there shouldn't be a warning. Maybe an info but even that may be stretching it.

> If I have a specification with a metakey default and the key does not exist in any of the namespaces, should it then automatically be created by the spec plugin?

Yes. Otherwise, the `default` metakey wouldn't be of much use, would it? Because if the key exists, the default value wouldn't be used.

> If I have a specification with a metakey require and the key does not exist in any of the namespaces, should it then return an error or also create the key by itself?

I think if there's no `default` then it should error. 

> In case 2. and 3. is answered with yes (both creating keys) should they be created in all namespaces?

No. That's what the `spec:/` namespace is for. See [elektra-namespaces](https://github.com/ElektraInitiative/libelektra/blob/master/doc/help/elektra-namespaces.md#spec):

> Usually, the spec-keys do not directly contribute to the value, with one notable exception: the default value (metadata default, see in cascading below) might be used if every other way as specified in the spec-key failed.

atmaxinger commented:
Well after digging through the old source code there might be another namespace for the default keys: `default:/`. Please create default keys in there.

markus2330 commented:
@atmaxinger thank you for your answers!

> I think if there's no default then it should error.

spec should only fail if required keys are missing (not just because default is missing)

>  In case 2. and 3. is answered with yes (both creating keys) should they be created in all namespaces?

As @atmaxinger said correctly: the default value is specified in spec: but the key to be created is in its own default: namespace (not to be persisted).

@tmakar please make sure that these details are also documented properly.

atmaxinger commented:
> spec should only fail if required keys are missing (not just because default is missing)

Yes, but the question was "what if there is `required` but no `default`". So then it should error. But if there's `required` and `default` I think it shouldn't.

tmakar commented:
@atmaxinger @markus2330 

Thank you both for your answers!! Helped a lot!

@markus2330 Yes ofc, I will document this, just wanted to clarify some things now 😉

tmakar commented:
FYI: @atmaxinger @markus2330 

**Implemented:**

- Meta data for non array specification and wildcards is copied to corresponding keys in other namespaces
- If default meta key is used in specification and is not present in configuration, key is created in `default` namespace
- If required meta key is specified and no configuration is present for it, `spec` sets error via one of the helper functions in `kdberrors.h` on the parent key
- If specification for key is there - and no `default` and `required` are present in specification for this key - an info is appended to the parent key (`meta:/info/description`).

For all those cases I created tests which verifying this behavior. For it I used helper functions from `tests_plugin.h`.

**TODO:**

- Implement array specification in `spec` (add validation for it, if the meta keys for validation are present)
- Implement wildcard (if array specification and wildcard collide add error)
- Extend documentation `README.md`
- Add description for `default` namespace in [Elektra Namespaces] Documentation`(https://www.libelektra.org/tutorials/namespaces)

Let me know if I forgot something!

Thanks again for your time!

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,225 @@
+	const Key * defaultMetaKey = keyGetMeta (specKey, "default");
+	const char * defaultValue = keyString (defaultMetaKey);
+
+	const char * parentKeyName = keyName (parentKey);
```
`keyName` will also include the namespace. You need to remove the namespace here, else you'll get `default:/system\:/blabla` if the parentkey is `system:/blabla`

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,225 @@
+ */
+static int copyMeta (Key * key, Key * specKey)
+{
+	KeySet * metaKeys = ksDup (keyMeta (specKey));
```
Why are you duplicating the returned KeySet?

tmakar replied:
You are right, should not be needed here.

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,225 @@
+static bool specMatches (Key * specKey, Key * otherKey)
+{
+	// ignore namespaces for globbing
+	Key * globKey = keyNew (strchr (keyName (otherKey), '/'), KEY_END);
```
See my comment about `keyName` below. It will contain the `otherKey`'s namespace.

atmaxinger replied:
You want to use `strchr (keyName (otherKey), '/')`

tmakar replied:
Already using it here, or do you mean something else?

tmakar commented:
**Implemented:**
- Extend and updated README.md
- Check for kdbGet => set warning if it is kdbGet
- Fix bug with appending namespace when creating key with default value in default namespace => add tests
- Use cascading lookup for checking `require`

**TODO:**
- Finish array specification instantiation and open TODOs (checks, array min / max)
- Wildcard specification
- Add `default:/` as namespace in Namespaces documentation

tmakar commented:
**Implemented:**
- Wildcard specification (one `_` in spec key supported)
- Extend tests
- Give tests a better description and some samples
- Add `default:/` as namespace in [namespaces.md](https://www.libelektra.org/tutorials/namespaces)

**TODO:**
- Support wildcard specification for more than one `_` in spec key
- Finish array specification instantiation and combine with metadata copying

tmakar commented:
@atmaxinger @hannes99 I have requested you for a review at this stage now, so you can review in smaller chunks. I have put my implementation progress in comments, so you can follow there.

FYI: @markus2330

markus2330 commented:
No need to mention me just because you assign someone for a review.

tmakar commented:
**Implemented:**
- Add more tests for wildcard specification implementation
- Add validation if array and wildcard specification have collision
- Add information about array and wildcard specification to `README` 

**TODO:**
- Finish array specification instantiation and combine with metadata copying
- Implement `elektraSpecGet`
- Implement `elektraSpecRemove`

tmakar commented:
**Implemented:**
- Array instantiation (add also helper functions)
- Replace custom `copyMeta` with `keyCopyAllMeta`

**TODO:**
- Fix bug with wildcard
- Implement `elektraSpecGet`
- Implement `elektraSpecRemove`
- Add more tests

tmakar commented:
@atmaxinger 
If I have the following configuration `user:/sw/org/server/name` and the specification `spec:/sw/org/server` with meta key `meta:/description`.

Should this meta key be then copied to `user:/sw/org/server` ?

tmakar commented:
**Implemented:**
- Added more tests
- Fixed bugs with array specification
- Added `elektraSpecGet`
- Added `ELEKTRA_PLUGIN_EXPORT`

**TODO:**
- Implement `elektraSpecRemove`

tmakar commented:
@atmaxinger @hannes99 @markus2330 

This is now ready to be reviewed. Implementation done, tests included and README is adapted.

Restyling will be fixed.

tmakar commented:
@atmaxinger 

Somehow this error is thrown in Jenkins:
```text
5 [ 2/11] RUN dnf config-manager --set-enabled powertools
#5 0.318 No such command: config-manager. Please use /usr/bin/dnf --help
#5 0.318 It could be a DNF plugin command, try: "dnf install 'dnf-command(config-manager)'"
#5 ERROR: process "/bin/sh -c dnf config-manager --set-enabled powertools" did not complete successfully: exit code: 1
```

Is this issue known ? 🤔

atmaxinger commented:
jenkins build libelektra please

atmaxinger commented:
> Is this issue known ? 🤔

Hmm it seems all of a sudden that the `dnf-plugins-core` package isn't installed anymore in the container.

atmaxinger commented:
Please rebase to master to fix the Jenkins issue

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+}
+
+/**
+ * Check if the specification key has a meta key `required`.
```
```suggestion
 * Check if the specification key has a meta key `require`.
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+/**
+ * Check if the specification key has a meta key `required`.
+ *
+ * @param specKey the specification key to check for the required meta key
```
```suggestion
 * @param specKey the specification key to check for the require meta key
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+ */
+static KeySet * extractSpecKeys (KeySet * ks)
+{
+	KeySet * specKeys = ksNew (0, KS_END);
```
```suggestion
	Key * specKey = keyNew ("spec:/", KEY_END);
	KeySet * specKeys = ksCut (ks, specKey);
```

Would be the entire function, no?

tmakar replied:
Yes, you are right!

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+}
+
+/**
+ * Checks if the {@link specKey} is a array specification key.
```
```suggestion
 * Checks if the {@link specKey} is an array specification key.
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+ */
+static bool isArraySpecification (Key * specKey)
+{
+	const char * keyWithoutNamespace = strchr (keyName (specKey), '/');
```
The whole function could just be 


```c
return strchr (keyName (specKey), '#') != NULL
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+
+	for (size_t i = 0; i < len; i++)
+	{
+		if (keyWithoutNamespace[i] == '#' && (i != len && keyWithoutNamespace[i + 1] == '_'))
```
`keyWithoutNamespace[i + 1] ` is a SIGSEGV waiting to happen.

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+ */
+static bool isWildcardSpecification (Key * specKey)
+{
+	const char * keyWithoutNamespace = strchr (keyName (specKey), '/');
```
This again could just be 

```c
return strchr (keyName (specKey), '_') != NULL;
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+ *
+ * @param arrayElement the element to store the name in
+ * @param arrayNumber the array element number
+ * @param size the size of array
```
```suggestion
 * @param size the allocated size of @p arrayElement
```

atmaxinger replied:
At least based on how you call this function later on. However, I'm not so sure now whether that's really what it is.

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+ * @param arrayNumber the array element number
+ * @param size the size of array
+ */
+static void createArrayElementName (char * arrayElement, int arrayNumber, int size)
```
Sorry, but either I am too stupid to understand what the function does, or the provided examples in the API doc are wrong.

```c
memset (buf, 0, 2048);
createArrayElementName (buf, 2, 2);
fprintf(stdout, "%s\n", buf);

memset (buf, 0, 2048);
createArrayElementName (buf, 10, 4);
fprintf(stdout, "%s\n", buf);

memset (buf, 0, 2048);
createArrayElementName (buf, 100, 6);
fprintf(stdout, "%s\n", buf);
```

Should output 
```
#2
#_10
#__100
``` 

But for me it outputs
```
#_2
#
#
```

Which makes total sense, as for `arrayNumber = 10` and `arrayNumber = 100` (or really anything multiple of 10) your loop will never execute.

atmaxinger replied:
I think this is what you actually want:


```c
static void createArrayElementName (char ** arrayElement, int arrayNumber, int size)
{
    // number of digits in arrayNumber
    int nDigits = floor(log10(abs(arrayNumber))) + 1;

    // allocate enough space for #, the digits and \0
    int minSize = nDigits + 1;
    size = size > minSize ? size : minSize;
    char * name = calloc ((size + 1) * sizeof (char));

    name[0] = '#';

    for (int j = 1; j < size - nDigits; j++)
    {
        name[j] = '_';
    }

    sprintf(&name[size-nDigits], "%d", arrayNumber);

    *arrayElement = name;
}
```

And then call it with:

```
char * arrayElementName = NULL;
createArrayElementName (&arrayElementName, i, /* whatever you want */);
elektraFree (arrayElementName);
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+	{
+		arrayElement[j] = '_';
+	}
+	sprintf (&arrayElement[size], "%d", arrayNumber);
```
The only time you call this function, you pass the number of allocated bytes for `arrayElement` as the parameter `size`. So the maximum index would be `size-1`. And here you want to write something _past_ that?

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+	char * copiedKeyName = elektraMalloc (elektraStrLen (specKeyName) + 1);
+	strcpy (copiedKeyName, (char *) specKeyName);
+
+	char * arrayParent = strtok (copiedKeyName, "#");
```
Please don't use `strtok`. It is not thread-safe. Use `strtok_r` or `strsep` instead.

tmakar replied:
Thanks!!!!!!

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+static void setArrayPositions (const char * keyNameWithoutNamespace, int * arrayPositions)
+{
+	int arrPos = 0;
+	for (size_t i = 0; i < elektraStrLen (keyNameWithoutNamespace); i++)
```
```suggestion
	for (int i = 0; i < elektraStrLen (keyNameWithoutNamespace); i++)
```

Just use `int` then you don't need to cast below.

tmakar replied:
Problem here is, that `elektraStrLen` returns `size_t`. I would need to cast it anyway, either once in the loop condition or each time on assignment. Probably just casting in the loop condition is "better" than doing it on each assignment, so I will do it there.

atmaxinger replied:
No you don't need to cast. The compiler will automatically upcast the `int` to `size_t` for the comparison.

tmakar replied:
Hmm, I get the following `error: comparison of integers of different signs: 'int' and 'size_t'`.

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+	{
+		if (keyNameWithoutNamespace[i] == '#')
+		{
+			arrayPositions[arrPos++] = (int) i;
```
```suggestion
			arrayPositions[arrPos++] = i;
```

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+		else
+		{
+			char * msg = elektraFormat ("Key for specification %s does not exist", keyName (specKey));
+			keySetMeta (parentKey, elektraFormat ("%s/%s", INFO_KEY, "description"), msg);
```
I think you also need to free the result of this `elektraFormat` call

atmaxinger commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,754 @@
+		// if required and no default => cascade lookup if exists in other namespaces
+		if ((isRequired (current) && !hasDefault (current)) || (isRequired (current) && isWildcardSpecification (current)))
+		{
+			Key * cascadingKey = keyNew (strchr (keyName (current), '/'), KEY_END);
```
You need to free this key somewhere

hannes99 reviewed (approved):
Looks fine IMHO

markus2330 reviewed (changes requested):
Great job, this is a huge improvement of the spec plugin's code quality.

tmakar commented on src/plugins/spec/CMakeLists.txt:
```diff
@@ -2,6 +2,6 @@ include (LibAddMacros)
-	SOURCES spec.h spec.c
-	LINK_ELEKTRA elektra-ease elektra-meta elektra-globbing
+	SOURCES spec-new.h spec-new.c
+	LINK_ELEKTRA elektra-ease elektra-meta elektra-globbing libm.so
```
yes, it is needed. What do you mean by proper CMake 🤔Currently `libm.so` is detected properly and links without any problems.

tmakar commented on src/plugins/spec/README.md:
```diff
@@ -179,11 +161,3 @@ kdb meta-ls spec:/freedesktop/openicc/device/camera/#0/EXIF_serial   # seems lik
-
-## Known Issues
-
-- In MINGW32 builds, there is no globbing and therefore no support for key names with # and \_.
```
> In MINGW32 builds, there is no globbing and therefore no support for key names with # and \_.

This is not working.

> Default values do not work if globbing is involved.

Default values are not allowed when globbing is used. IMO this is an acceptable tradeoff.

> Added metadata is not correctly removed during `kdbSet`, if the corresponding spec key was modified.

I will verify this.

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,758 @@
+ * @param arrayNumber the array element number
+ * @return the created array element e.g. #0, #_10, #_100
+ */
+static char * createArrayElementName (int arrayNumber)
```
Hmm, I couldn't find any :/

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,758 @@
+}
+
+/**
+ * Appends a key to the KeySet with the default value specified in the meta key (meta:/default) in the
```
Thanks 😃

flo91 commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+ * The positions of where the `#` occurs in the specification key.
+ *
+ * @param keyNameWithoutNamespace the key name without the namespace
+ * @param arrayPositions the array to hold all the positions of the `#` in the specification key
```
```suggestion
 * @param[out] arrayPositions the array to hold all the positions of the `#` in the specification key
```

flo91 commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+{
+	Key * key = ksLookupByName (ks, strchr (keyName (specKey), '/'), 0);
+	Key * keyToFetchArraySizeFrom = key == NULL ? getArraySizeOfArrayParent (specKeys, specKey) : key;
+	const Key * arrayMetaKey = key == NULL ? keyGetMeta (keyToFetchArraySizeFrom, "array") : keyGetMeta (key, "array");
```
```suggestion
	const Key * arrayMetaKey = keyGetMeta (keyToFetchArraySizeFrom, "array");
```
Isn't this redundant?
If `key != NULL`, `keyToFetchArraySizeFrom` is set to `key`, so it's enough to always use `keyToFetchArraySizeFrom` here.

tmakar replied:
Yes, ofc, thanks!

flo91 commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+		return true;
+	}
+
+	if (!validateArraySize (keyToFetchArraySizeFrom, specKey))
+	{
+		return false;
+	}
+
+	return true;
```
Could be simplified.

```suggestion
	return validateArraySize (keyToFetchArraySizeFrom, specKey);
```

flo91 commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+		char * withoutNamespace = strchr (keyName (current), '/');
+
+		int len = elektraStrLen (withoutNamespace);
+		if (arrayPosition > len)
```
```suggestion
		if (arrayPosition >= (len-1))
```
`len` are the numbers of characters including `\0`, so the last valid index is `len-1` where the `\0` has to be.
The last possible position for the array character is therefore at `len-2`.

atmaxinger replied:
Isn't `> len` equal to `>= len-1`?

tmakar replied:
@atmaxinger The other way around. If `>= len+1`, just `> len` would be fine.

flo91 commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+		}
+
+		if (withoutNamespace[arrayPosition] == '#' &&
+		    (arrayPosition < len && (withoutNamespace[arrayPosition + 1] == '_' || withoutNamespace[arrayPosition + 1] != '/' ||
```
```suggestion
		    (arrayPosition < (len-1) && (withoutNamespace[arrayPosition + 1] == '_' || withoutNamespace[arrayPosition + 1] != '/' ||
```

flo91 commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+ * be created.
+ *
+ * In case the key is missing and has no meta key default the method returns -1 and it fails by
+ * adding error.
```
```suggestion
 * adding an error (or warning in the case of `kdbGet`).
```

flo91 reviewed (changes requested):
Nice progress and tests! 😃 

However, I've still found some bugs and errors, esp. regarding memory management and array bounds. Please make sure to fix them carefully.

tmakar commented on src/plugins/spec/README.md:
```diff
@@ -37,106 +33,97 @@ Specifically it supports:
-`array/max` have to be valid array-elements similar to `array`. If the array size is out of bounds, this causes a `range` conflict.
+set, we validate the array size (given as metakey `array` on `x`) is within the limits of `array/min` and `array/max`.
+Both `array/min` and `array/max` have to be valid array-elements similar to `array`.
+If the array size is out of bounds, this causes an `error`.
+If it is a `kdbGet` it is a `warning`.
```
Yes, exactly, thanks!

tmakar commented on src/plugins/spec/testmod_spec.c:
```diff
@@ -1,655 +1,702 @@
- *
- * @brief
- *
- * @copyright BSD License (see doc/LICENSE.md or https://www.libelektra.org)
```
No, will put it in again.

tmakar commented on src/plugins/spec/testmod_spec.c:
```diff
@@ -1,655 +1,702 @@
+	}
 
-		TEST_CHECK (elektraSpecCopy (plugin, ks, parentKey, isKdbGet) == ELEKTRA_PLUGIN_STATUS_ERROR, "hook spec/copy should fail");
+	const Key * infoDescription = keyGetMeta (parentKey, elektraFormat ("%s/%s", INFO_KEY, "description"));
+	printf ("info with description on key %s: %s\n", keyName (parentKey), keyString (infoDescription));
 
```
Yes, thanks have forgotten to fix this in tests as well.

tmakar commented on src/plugins/spec/testmod_spec.c:
```diff
@@ -1,655 +1,702 @@
+				succeed_if_same_string (keyName (current), keyNameToMatch)
+			}
+		}
+
```
Thanks, forgot it here in tests as well.

tmakar commented on src/plugins/spec/testmod_spec.c:
```diff
@@ -1,655 +1,702 @@
+				{
+					count++;
+				}
+				succeed_if (keyGetNamespace (current) == KEY_NS_DEFAULT, "should not contain default");
```
Yes, exactly!

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+ * The {@link newStr} contains the string with the replaced character.
+ *
+ * @param str the old string
+ * @param newStr the newly created string with the already replaced character
```
Thanks. Yes, for this function it makes definitely sense to explicitly state the out-parameter.

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+ */
+static bool containsUnderlineInArraySpec (Key * specKey)
+{
+	return strchr (keyName (specKey), '_') != NULL;
```
No, this is done in a separate function. I wanted to have two different function and each having its own purpose.

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+		name[j] = '_';
+	}
+
+	sprintf (&name[nDigits], "%d", arrayNumber);
```
Yes, you are completely right. This is a mistake. Should be allocated for `nDigits + 2`.

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+{
+	char * strUntilArrayElement = elektraMalloc (pos);
+	memcpy (strUntilArrayElement, &keyNameWithoutNamespace[0], pos - 1);
+	strUntilArrayElement[pos] = '\0';
```
I changed it now to allocate `pos+1` memory.

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+	strUntilArrayElement[pos] = '\0';
+
+	size_t keyNameSize = elektraStrLen (keyNameWithoutNamespace);
+	char * strAfterArrayElement = elektraMalloc (keyNameSize + 1);
```
Ah, I see, thanks.

tmakar commented on src/plugins/spec/spec-new.c:
```diff
@@ -0,0 +1,762 @@
+	memcpy (strUntilArrayElement, &keyNameWithoutNamespace[0], pos - 1);
+	strUntilArrayElement[pos] = '\0';
+
+	size_t keyNameSize = elektraStrLen (keyNameWithoutNamespace);
```
Yes, but I still need the size of the whole keyName, for copying the right substring from `keyNameWithoutNamespace`.

tmakar commented:
@markus2330 @atmaxinger @flo91 

Thank you again for the time you already invested in reviewing!!!!

tmakar commented:
@flo91 

Do you maybe know what this error is about

```sh
Sorry, 3 warnings were issued ;(
 1: Module 
kdb
 issued the warning 
C01320
:
	Interface: Calling the kdbGet function for the backend plugin ('backend') of the mountpoint 'system:/elektra' has failed during the STORAGE phase.
 2: Module 
kdb
 issued the warning 
C01320
:
	Interface: The STORAGE phase of kdbGet() has failed. See warnings for details.
 3: Module 
kdb
 issued the warning 
C03100
:
	Validation Syntactic: Expected newline '\n' but got 'n' at position 4000.
```

The `dump` storage plugin seems to return `Expected newline '\n' but got 'n' at position 4000`.

tmakar commented:
```test
[100%] Built target application
Mountpoint spec:/sw/example/highlevel/#0/current does not exist
Mountpoint /sw/example/highlevel/#0/current does not exist
Did not find any key
Did not find any key
successfully read configuration
Create a new key user:/sw/example/highlevel/#0/current/mystring with string "Hello World"
Create a new key user:/sw/example/highlevel/#0/current/myint with string "29"
Create a new key user:/sw/example/highlevel/#0/current/mydouble with string "4.4242"
Create a new key user:/sw/example/highlevel/#0/current/myfloatarray/#0 with string "3.14"
Create a new key user:/sw/example/highlevel/#0/current/myfloatarray/#1 with string "15"
Create a new key user:/sw/example/highlevel/#0/current/myfloatarray/#2 with string "9265.359"
Create a new key user:/sw/example/highlevel/#0/current/myfloatarray/#3 with string "2.718282"
Create a new key user:/sw/example/highlevel/#0/current/myfloatarray/#4 with string "1.4142"
ERROR: The key '/sw/example/highlevel/#0/current/mystring' has the wrong type (expected 'string' but got '(null)').
error: application could not read changed config
Create a new key user:/sw/example/highlevel/#0/current/print with string "1"
ERROR: The key '/sw/example/highlevel/#0/current/mystring' has the wrong type (expected 'string' but got '(null)').
error: application could not read changed config
```

@atmaxinger @flo91 Is there something I missed? Is the `spec` plugin called on `key` creation or why do I run into a type mismatch here?

atmaxinger commented:
> The dump storage plugin seems to return Expected newline '\n' but got 'n' at position 4000.

Do you have access to the storage file?

atmaxinger commented:
> Is the spec plugin called on key creation

No, the spec plugin is only called during `kdbGet` and `kdbSet`. AFAIK there is no validation at all outside of that.

tmakar commented:
> > The dump storage plugin seems to return Expected newline '\n' but got 'n' at position 4000.
> 
> Do you have access to the storage file?

It's run inside of the `cirrus ci` jobs, I don't know. This is the information I extracted from the logs of this ci jobs.

tmakar commented:
> > Is the spec plugin called on key creation
> 
> No, the spec plugin is only called during `kdbGet` and `kdbSet`. AFAIK there is no validation at all outside of that.

Hmm :/ This is really weird.

tmakar commented:
@atmaxinger when `kdb set <namespace:/some-key-name> <value>` is run, I assume `kdbSet` is called, right?

tmakar commented:
And another question, what validation are we talking about?

tmakar commented:
@atmaxinger @markus2330 @flo91 

Currently 99% of the tests are passing. There is still a problem with 2-3 other tests.

1. check_external_example_codegen_econf
2. check_external_example_codegen_menu
3. check_external_example_highlevel

I investigated this, but it seems to be a really weird issue. 

The error seems to be always the same, an example is below.

```sh
ERROR: The key '/sw/example/menu/#0/current/menu/#1/command' has the wrong type (expected 'string' but got '(null)').
error: application could not read test menu
```

I am not sure why the `type` is `(null)`. I have a lot of tests in `testmod_spec.c` an all are copying the `type` correctly. Am I missing some implementation detail? 
Do I need to adapt anything in those failing tests? If yes, what and where?

I would be happy if you could help me out.

Open tasks:
- Restructure some functions into separate files
- Adapt release notes (make it a highlight)

tmakar commented:
**Done:**
- Make rewrite spec plugin a highlight in release notes
- Extract array spec functions into separate file

**TODO:**
- Add shell recorder samples to `README`

As already mentioned above, 99% of tests pass, but there are 3 which don't.

tmakar commented:
**Done:**
- Added specification example to readme with shell_recorder test and explanation

markus2330 reviewed (changes requested):
Please fix remaining tests first.

tmakar commented:
@atmaxinger @markus2330 

I ran the test `check_external_example_highlevel` locally now.

All the tests have passed.

Here is the output of the tests:

[example_highlevel_test_new_spec_branch_output.txt](https://github.com/ElektraInitiative/libelektra/files/11202611/example_highlevel_test_new_spec_branch_output.txt)

tmakar commented:
On `FreeBSD 13` I get the following:

I don't know how this is related to my changes.

```sh
error: application did not read test menu correctly
 Sorry, 3 warnings were issued ;(
 1: Module 
kdb
 issued the warning 
C01320
:
	Interface: Calling the kdbGet function for the backend plugin ('backend') of the mountpoint 'system:/elektra' has failed during the STORAGE phase.
 2: Module 
kdb
 issued the warning 
C01320
:
	Interface: The STORAGE phase of kdbGet() has failed. See warnings for details.
 3: Module 
kdb
 issued the warning 
C03100
:
	Validation Syntactic: Expected newline '\n' but got 'n' at position 4000.
```

tmakar commented:
@markus2330 @atmaxinger @flo91 @hannes99 

- All tests are now passing! :heavy_check_mark: 
- All review comments have been implemented :heavy_check_mark: 
- Implementation (Rewrite) is completely finished :heavy_check_mark: 
- Tests added for implementation :heavy_check_mark: 
- Shell recorder test added to README :heavy_check_mark: 
- Adapted README :heavy_check_mark: 
- Added `default` namespace to namespaces tutorial :heavy_check_mark: 

This is now ready for a final review round!!!!
If everything is 

I already want to thank for your time :wink: 

Would be cool if we could merge this next week!!

atmaxinger reviewed (approved):
Great work! I've tested it with the [opensesame specification](https://github.com/ElektraInitiative/opensesame/blob/master/files/opensesame.spec) and it worked marvellously.

I have, however, found one subtle bug, but not in your code. In `kdbSet` we do not necessarily call the spec plugin with the `spec:/` keys in the `returned` keyset.  This leads to the plugin not removing metadata. 

This is a bug in `kdbSet`. I've fixed that, here is the diff:

```diff
diff --git a/src/libs/elektra/kdb.c b/src/libs/elektra/kdb.c
index 5d53fe652..30dcdb9cb 100644
--- a/src/libs/elektra/kdb.c
+++ b/src/libs/elektra/kdb.c
@@ -2496,8 +2496,14 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
        ksClear (setKs);
        backendsMerge (backends, setKs);
 
+       // It's not guaranteed that the spec:/ keys are included in setKs
+       // Therefore we copy the spec keys from ks and append the keys from setKs
+       Key * specRoot = keyNew ("spec:/", KEY_END);
+       KeySet * withSpecKeys = ksBelow (ks, specRoot);
+       ksAppend (withSpecKeys, setKs);
+
        // Step 9: run the spec plugin to remove copied metadata
-       if (handle->hooks.spec.plugin && handle->hooks.spec.remove (handle->hooks.spec.plugin, setKs, parentKey) == -1)
+       if (handle->hooks.spec.plugin && handle->hooks.spec.remove (handle->hooks.spec.plugin, withSpecKeys, parentKey) == -1)
        {
                goto rollback;
        }
@@ -2581,6 +2587,9 @@ error:
        ksDel (setKs);
        ksDel (backends);
 
+       ksDel (withSpecKeys);
+       keyDel (specRoot);
+
        errno = errnosave;
 
        return -1;
```

Please go ahead an apply this diff on your code base. Then I think this is ready for merging.

tmakar commented:
> Great work! I've tested it with the [opensesame specification](https://github.com/ElektraInitiative/opensesame/blob/master/files/opensesame.spec) and it worked marvellously.
> 
> I have, however, found one subtle bug, but not in your code. In `kdbSet` we do not necessarily call the spec plugin with the `spec:/` keys in the `returned` keyset. This leads to the plugin not removing metadata.
> 
> This is a bug in `kdbSet`. I've fixed that, here is the diff:
> 
> ```diff
> diff --git a/src/libs/elektra/kdb.c b/src/libs/elektra/kdb.c
> index 5d53fe652..30dcdb9cb 100644
> --- a/src/libs/elektra/kdb.c
> +++ b/src/libs/elektra/kdb.c
> @@ -2496,8 +2496,14 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
>         ksClear (setKs);
>         backendsMerge (backends, setKs);
>  
> +       // It's not guaranteed that the spec:/ keys are included in setKs
> +       // Therefore we copy the spec keys from ks and append the keys from setKs
> +       Key * specRoot = keyNew ("spec:/", KEY_END);
> +       KeySet * withSpecKeys = ksBelow (ks, specRoot);
> +       ksAppend (withSpecKeys, setKs);
> +
>         // Step 9: run the spec plugin to remove copied metadata
> -       if (handle->hooks.spec.plugin && handle->hooks.spec.remove (handle->hooks.spec.plugin, setKs, parentKey) == -1)
> +       if (handle->hooks.spec.plugin && handle->hooks.spec.remove (handle->hooks.spec.plugin, withSpecKeys, parentKey) == -1)
>         {
>                 goto rollback;
>         }
> @@ -2581,6 +2587,9 @@ error:
>         ksDel (setKs);
>         ksDel (backends);
>  
> +       ksDel (withSpecKeys);
> +       keyDel (specRoot);
> +
>         errno = errnosave;
>  
>         return -1;
> ```
> 
> Please go ahead an apply this diff on your code base. Then I think this is ready for merging.

Thank you!!!!!!

tmakar commented:
jenkins build libelektra please

markus2330 commented:
Great job! Will test it when packages are build for it!
